File: src/test/java/no/nav/fo/veilarbregistrering/enhet/adapter/EnhetGatewayTest.java
package no.nav.fo.veilarbregistrering.enhet.adapter;

import no.nav.fo.veilarbregistrering.arbeidsforhold.Organisasjonsnummer;
import no.nav.fo.veilarbregistrering.enhet.Kommunenummer;
import no.nav.fo.veilarbregistrering.enhet.Organisasjonsdetaljer;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import javax.ws.rs.NotFoundException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class EnhetGatewayTest {

    private EnhetGatewayImpl enhetGateway;

    @BeforeEach
    public void setUp() {
        EnhetRestClient enhetStubClient = new EnhetStubClient();
        enhetGateway = new EnhetGatewayImpl(enhetStubClient);
    }

    @Test
    public void hentOrganisasjonsdetaljer_skal_kunne_hente_ut_kommunenummer_fra_enhetsregisteret() {
        Optional<Organisasjonsdetaljer> organisasjonsdetaljer
                = enhetGateway.hentOrganisasjonsdetaljer(Organisasjonsnummer.of("995298775"));

        assertThat(organisasjonsdetaljer).isNotEmpty();
        assertThat(organisasjonsdetaljer.get().kommunenummer()).hasValue(Kommunenummer.of("0301"));
    }

    @Test
    public void hentOrganisasjonsdetaljer_skal_gi_empty_result_ved_ukjent_org_nr() {
        Optional<Organisasjonsdetaljer> organisasjonsdetaljer
                = enhetGateway.hentOrganisasjonsdetaljer(Organisasjonsnummer.of("123456789"));

        assertThat(organisasjonsdetaljer).isEmpty();
    }

    @Test
    public void hentOrganisasjonsdetaljer_skal_kaste_runtimeException_ved_feil() {
        RuntimeException runtimeException = assertThrows(RuntimeException.class,
                () -> enhetGateway.hentOrganisasjonsdetaljer(null));

        assertThat(runtimeException.getMessage()).isEqualTo("Hent organisasjon feilet");
    }

    private class EnhetStubClient extends EnhetRestClient {

        private static final String OK_JSON = "/enhet/enhet.json";

        private Map<Organisasjonsnummer, String> jsonResponse = new HashMap<>();

        EnhetStubClient() {
            super(null);
            jsonResponse.put(Organisasjonsnummer.of("995298775"), toJson(OK_JSON));
        }

        @Override
        String utfoerRequest(Organisasjonsnummer organisasjonsnummer) {

            if (organisasjonsnummer == null) {
                throw new RuntimeException("Stub: Hent organisasjon feilet");
            }

            String jsonResponse = this.jsonResponse.get(organisasjonsnummer);
            if (jsonResponse == null) {
                throw new NotFoundException("Stub: Fant ikke organisasjonsnummer");
            }
            return jsonResponse;
        }
    }

    private String toJson(String json_file) {
        try {
            byte[] bytes = Files.readAllBytes(Paths.get(EnhetRestClient.class.getResource(json_file).toURI()));
            return new String(bytes);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/test/java/no/nav/fo/veilarbregistrering/enhet/adapter/EnhetGatewayTest.kt
package no.nav.fo.veilarbregistrering.enhet.adapter

import no.nav.fo.veilarbregistrering.FileToJson
import no.nav.fo.veilarbregistrering.arbeidsforhold.Organisasjonsnummer
import no.nav.fo.veilarbregistrering.arbeidsforhold.Organisasjonsnummer.Companion.of
import no.nav.fo.veilarbregistrering.enhet.Kommunenummer
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockserver.integration.ClientAndServer
import org.mockserver.junit.jupiter.MockServerExtension
import org.mockserver.model.HttpRequest
import org.mockserver.model.HttpResponse
import org.mockserver.model.MediaType

@ExtendWith(MockServerExtension::class)
class EnhetGatewayTest(private val mockServer: ClientAndServer) {
    private lateinit var enhetGateway: EnhetGatewayImpl

    @BeforeEach
    fun setUp() {
        val baseUrl = "http://" + mockServer.remoteAddress().address.hostName + ":" + mockServer.remoteAddress().port
        enhetGateway = EnhetGatewayImpl(EnhetRestClient(baseUrl))
    }

    @Test
    fun `hentOrganisasjonsdetaljer skal kunne hente ut kommunenummer fra enhetsregisteret`() {
        val org = of("995298775")
        mockServer.`when`(
            HttpRequest
                .request()
                .withMethod("GET")
                .withPath("/api/v1/organisasjon/${org.asString()}"))
            .respond(
                HttpResponse.response()
                .withStatusCode(200)
                .withBody(OK_JSON, MediaType.JSON_UTF_8))

        val organisasjonsdetaljer = enhetGateway.hentOrganisasjonsdetaljer(org)
        assertThat(organisasjonsdetaljer).isNotNull
        assertThat(organisasjonsdetaljer!!.kommunenummer()).hasValue(Kommunenummer("0301"))
    }

    @Test
    fun `hentOrganisasjonsdetaljer skal gi empty result ved ukjent org nr`() {
        val ukjentOrg = of("123456789")

        mockServer.`when`(
            HttpRequest
                .request()
                .withMethod("GET")
                .withPath("/api/v1/organisasjon/${ukjentOrg.asString()}"))
            .respond(
                HttpResponse.response()
                    .withStatusCode(404))

        val organisasjonsdetaljer = enhetGateway.hentOrganisasjonsdetaljer(ukjentOrg)
        assertThat(organisasjonsdetaljer).isNull()
    }

    @Test
    fun `hentOrganisasjonsdetaljer skal kaste runtimeException ved feil`() {
        mockServer.`when`(
            HttpRequest
                .request()
                .withMethod("GET")
                .withPath("/api/v1/organisasjon/${FEILENDE_ORG.asString()}"))
            .respond(
                HttpResponse.response()
                    .withStatusCode(500))

        val runtimeException = assertThrows<RuntimeException> { enhetGateway.hentOrganisasjonsdetaljer(FEILENDE_ORG) }
        assertThat(runtimeException.message).isEqualTo("Hent organisasjon feilet med status: 500")
    }

    companion object {
        private val OK_JSON = FileToJson.toJson("/enhet/enhet.json")
        private val FEILENDE_ORG: Organisasjonsnummer = of("0")
    }
}
